File: python/task_list/task_list.py
from python.task_list.task import Task


class TaskList:
    QUIT = "quit"

    def __init__(self, console):
        self.last_id = 0
        self.console = console
        self.tasks = dict()

    def run(self):
        while True:
            command = self.console.input("> ")
            if command == self.QUIT:
                break
            self.execute(command)

    def execute(self, command_line):
        command_rest = command_line.split(" ", 1)
        command = command_rest[0]
        if command == "show":
            self.show()
        elif command == "add":
            self.add(command_rest[1])
        elif command == "check":
            self.check(command_rest[1])
        elif command == "uncheck":
            self.uncheck(command_rest[1])
        elif command == "help":
            self.help()
        else:
            self.error(command)

    def show(self):
        for project, tasks in self.tasks.items():
            self.console.print(project)
            for task in tasks:
                self.console.print(f"  [{'x' if task.is_done() else ' '}] {task.id}: {task.description}")
            self.console.print()

    def add(self, command_line):
        sub_command_rest = command_line.split(" ", 1)
        sub_command = sub_command_rest[0]
        if sub_command == "project":
            self.add_project(sub_command_rest[1])
        elif sub_command == "task":
            project_task = sub_command_rest[1].split(" ", 1)
            self.add_task(project_task[0], project_task[1])

    def add_project(self, name):
        self.tasks[name] = []

    def add_task(self, project, description):
        project_tasks = self.tasks.get(project)
        if project_tasks is None:
            self.console.print(f"Could not find a project with the name {project}.")
            self.console.print()
            return
        project_tasks.append(Task(self.next_id(), description, False))

    def error(self, command):
        self.console.print(f"I don't know what the command {command} is.")
        self.console.print()

    def next_id(self):
        self.last_id += 1
        return self.last_id

    def check(self, id_string):
        self.set_done(id_string, True)

    def uncheck(self, id_string):
        self.set_done(id_string, False)

    def set_done(self, id_string, done):
        id_ = int(id_string)
        for project, tasks in self.tasks.items():
            for task in tasks:
                if task.id == id_:
                    task.set_done(done)
                    return
        self.console.print(f"Could not find a task with an ID of {id_}")
        self.console.print()

    def help(self):
        self.console.print("Commands:")
        self.console.print("  show")
        self.console.print("  add project <project name>")
        self.console.print("  add task <project name> <task description>")
        self.console.print("  check <task ID>")
        self.console.print("  uncheck <task ID>")
        self.console.print()

File: kotlin/src/main/java/com/codurance/training/tasks/TaskList.kt
package com.codurance.training.tasks

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.PrintWriter
import java.util.*

class TaskList(private val `in`: BufferedReader, private val out: PrintWriter) : Runnable {

    private val tasks = LinkedHashMap<String, MutableList<Task>>()

    private var lastId: Long = 0

    override fun run() {
        while (true) {
            out.print("> ")
            out.flush()
            val command: String
            try {
                command = `in`.readLine()
            } catch (e: IOException) {
                throw RuntimeException(e)
            }

            if (command == QUIT) {
                break
            }
            execute(command)
        }
    }

    private fun execute(commandLine: String) {
        val commandRest = commandLine.split(" ".toRegex(), 2).toTypedArray()
        val command = commandRest[0]
        when (command) {
            "show" -> show()
            "add" -> add(commandRest[1])
            "check" -> check(commandRest[1])
            "uncheck" -> uncheck(commandRest[1])
            "help" -> help()
            else -> error(command)
        }
    }

    private fun show() {
        for ((key, value) in tasks) {
            out.println(key)
            for (task in value) {
                out.printf("    [%c] %d: %s%n", if (task.isDone) 'x' else ' ', task.id, task.description)
            }
            out.println()
        }
    }

    private fun add(commandLine: String) {
        val subcommandRest = commandLine.split(" ".toRegex(), 2).toTypedArray()
        val subcommand = subcommandRest[0]
        if (subcommand == "project") {
            addProject(subcommandRest[1])
        } else if (subcommand == "task") {
            val projectTask = subcommandRest[1].split(" ".toRegex(), 2).toTypedArray()
            addTask(projectTask[0], projectTask[1])
        }
    }

    private fun addProject(name: String) {
        tasks[name] = ArrayList()
    }

    private fun addTask(project: String, description: String) {
        val projectTasks = tasks[project]
        if (projectTasks == null) {
            out.printf("Could not find a project with the name \"%s\".", project)
            out.println()
            return
        }
        projectTasks.add(Task(nextId(), description, false))
    }

    private fun check(idString: String) {
        setDone(idString, true)
    }

    private fun uncheck(idString: String) {
        setDone(idString, false)
    }

    private fun setDone(idString: String, done: Boolean) {
        val id = Integer.parseInt(idString)
        for ((_, value) in tasks) {
            for (task in value) {
                if (task.id == id.toLong()) {
                    task.isDone = done
                    return
                }
            }
        }
        out.printf("Could not find a task with an ID of %d.", id)
        out.println()
    }

    private fun help() {
        out.println("Commands:")
        out.println("  show")
        out.println("  add project <project name>")
        out.println("  add task <project name> <task description>")
        out.println("  check <task ID>")
        out.println("  uncheck <task ID>")
        out.println()
    }

    private fun error(command: String) {
        out.printf("I don't know what the command \"%s\" is.", command)
        out.println()
    }

    private fun nextId(): Long {
        return ++lastId
    }

    companion object {
        private val QUIT = "quit"

        @Throws(Exception::class)
        @JvmStatic
        fun main(args: Array<String>) {
            val `in` = BufferedReader(InputStreamReader(System.`in`))
            val out = PrintWriter(System.out)
            TaskList(`in`, out).run()
        }
    }
}
